File: ProjectDemo/SinaWBFrameModel/Classes/SinaPictureView.m
//
//  SinaPictureView.m
//  FrameModel方式
//
//  Created by san_xu on 2017/3/17.
//  Copyright © 2017年 朱献国. All rights reserved.
//

#import "SinaPictureView.h"
#import "UIImageView+WebCache.h"

static CGFloat const ItemMargin = 5.0;
#define ItemWitdh ((SCREENW - 2 * MARGIN - 2 * ItemMargin) / 3)
static NSString * const reuse = @"SinaPictureView";

@interface SinaPictureView ()<UICollectionViewDataSource>

@end

@implementation SinaPictureView

#pragma mark -API
+ (instancetype)sinaPictureView {
    
    UICollectionViewFlowLayout *flowLayout = [[UICollectionViewFlowLayout alloc] init];
    SinaPictureView * pictureView = [[SinaPictureView alloc] initWithFrame:CGRectZero collectionViewLayout:flowLayout];
    
    flowLayout.itemSize = CGSizeMake(ItemWitdh, ItemWitdh);
    flowLayout.minimumLineSpacing = ItemMargin;
    flowLayout.minimumInteritemSpacing = ItemMargin;
    
    pictureView.dataSource = pictureView;
    pictureView.backgroundColor = [UIColor whiteColor];
    [pictureView registerClass:[UICollectionViewCell class] forCellWithReuseIdentifier:reuse];
    
    
    return pictureView;
}

+ (CGSize)picSize:(NSInteger)picCount {
    
    if (picCount == 1) {
        return  CGSizeMake(ItemWitdh, ItemWitdh);
    } else if (picCount == 4) {
        return CGSizeMake(2 * ItemWitdh + ItemMargin, 2 * ItemWitdh + ItemMargin);
    } else {
        NSInteger count = (picCount - 1) / 3 + 1;
        CGFloat height = ItemWitdh * count + (count - 1) * ItemMargin;
        return CGSizeMake(SCREENW - 2 * MARGIN, height);
    }
  
}

- (void)setPicArr:(NSArray *)picArr {
    _picArr = picArr;
    
    //
    [self reloadData];
}


#pragma mark - UICollectionView DataSource

- (NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section {
    return self.picArr.count;
}

- (UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath {
    
    UICollectionViewCell *cell = [collectionView dequeueReusableCellWithReuseIdentifier:reuse forIndexPath:indexPath];
    
    //添加imageview
    UIImageView *img = [[UIImageView alloc] init];
    [cell addSubview:img];
    img.frame = cell.bounds;
    
    //
    NSString *imgPath = self.picArr[indexPath.item][@"thumbnail_pic"];
    [img sd_setImageWithURL:[NSURL URLWithString:imgPath] placeholderImage:[UIImage imageNamed:@"timeline_image_placeholder"]];
    
    return cell;
}
//- (CGSize)picSizeWithCount:(NSInteger)picCount {
//    
//    
//    return 0;
//}

@end

File: ProjectDemo/SinaWBAutoCal/Class/SinaPictureView.swift
//
//  SinaPictureView.swift
//  SinaSwiftLearn
//
//  Created by san_xu on 2017/3/15.
//  Copyright © 2017年 huakala. All rights reserved.
//

import UIKit

let SinaPictureViewCell = "SinaPictureViewCell"

class SinaPictureView: UICollectionView,UICollectionViewDataSource {
    
    // 定义一个变量
    var pics: [[String:Any]]? {

        didSet{
            //获取Size
            let size = getSize()
            self.snp.updateConstraints { (make) in
                make.width.equalTo(size.width)
                make.height.equalTo(size.height)
            }
            //刷新
            reloadData()
        }
        
    }
    
    let flowLayout = UICollectionViewFlowLayout()//布局对象
    init() {
        super.init(frame:CGRect(x: 0, y: 0, width: 0, height: 0) , collectionViewLayout: flowLayout)
        setupSubview()
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    private func setupSubview() {
        backgroundColor = UIColor.white
        
        //设置数据源
        dataSource = self
        
        //设置大小
        flowLayout.itemSize = CGSize(width: ItemWidth, height: ItemWidth)
        flowLayout.minimumLineSpacing = ItemMargin
        flowLayout.minimumInteritemSpacing = ItemMargin
        
        //注册cell
        self.register(UICollectionViewCell.self, forCellWithReuseIdentifier: SinaPictureViewCell)
    }
    
    func getSize() -> CGSize {
        
        if pics?.count == 1 {
            return CGSize(width: ItemWidth, height: ItemWidth)
        } else if pics?.count == 4 {
            return CGSize(width: (2 * ItemWidth + ItemMargin), height: (2 * ItemWidth + ItemMargin))
        } else {
            if let imgCount = pics?.count {
                let row = (imgCount - 1 ) / 3 + 1
                let heigth = CGFloat(row) * (ItemWidth) + (CGFloat(row) - 1) * ItemMargin
                return CGSize(width: SCREENW - CGFloat(2 * MARGIN), height: heigth)
            }
        }
        return CGSize()
    }

}

extension SinaPictureView {
    
    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        return pics?.count ?? 0
    }
    
    @objc(collectionView:cellForItemAtIndexPath:) func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
        
        let cell = collectionView.dequeueReusableCell(withReuseIdentifier: SinaPictureViewCell, for: indexPath)
        
        let imgView = UIImageView()
        cell.contentView.addSubview(imgView)
        imgView.frame = cell.bounds
        let dict = pics?[indexPath.item]
        imgView.sd_setImage(with: URL(string: dict?["thumbnail_pic"] as! String))
        
        return cell
        
    }
    
}
